### PR TITLE
[Entitlements] Consider only system modules in the boot layer

### PR DESCRIPTION
**Pull request by ldematte**
The JDK methods we instrument with _entitlement checks_ call an `EntitlementChecker`. This `EntitlementChecker` needs to establish if the caller is entitled to do some action. The production implementation of Entitlements does that by using modules: the `EntitlementChecker` starts with the module of the caller class and look up for the "requesting module" in the call chain. The "requesting module" is the first non-JDK module (typically, and ES module); for that, we want to filter out JDK (system) modules that may stand in between (e.g. JDK method calling another JDK method).

Currently we filter out all modules in the boot layer; this is incorrect, as everything on the module path (i.e. everything directly in the ES lib dir, which is server and its dependencies) is in the boot layer.

This PR adds a map of modules that are in the boot layer _and_ are system modules (i.e. the modules in the Java run-time image) as located by the system module finder `ModuleFinder.ofSystem()`. This map is then used to check if the module in the call chain is a system module or not.

### CODE REVIEW

**Review by prdoyle — APPROVED** [review]
Looks great. No notes!

File: libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java [hunk 1]
```diff
@@ -23,20 +27,46 @@
...
+            .map(ModuleReference::descriptor)
+            .collect(Collectors.toUnmodifiableSet());
+
+        var systemModules = ModuleLayer.boot()
+            .modules()
+            .stream()
+            .filter(m -> systemModulesDescriptors.contains(m.getDescriptor()))
+            .collect(Collectors.toUnmodifiableSet());
+
+        logger.info("System modules: " + systemModules);
```
Comment (prdoyle) [inline libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java:45]: Is the `info` level right here? We might want this as a `debug`.

File: libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java [hunk 2]
```diff
@@ -61,10 +91,6 @@ private static boolean isTriviallyAllowed(Module requestingModule) {
             logger.debug("Trivially allowed: Entire call stack is in the boot module layer");
```
Comment (prdoyle) [inline libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java:91]: This log message doesn't seem right anymore. It's more like the entire call stack is composed of system modules?

**Review by prdoyle — COMMENTED** [review]
On second thought, a couple of notes about the logging.

File: libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java (hunk 1 above)
Comment (ldematte) [reply libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java:45]: That's definitely a leftover from testing, I'll remove it, thanks!

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**